{
 "metadata": {
  "name": "",
  "signature": "sha256:9b60bc6bf87709a6852e73d4e551730f2f779e9913fd7581897549c9cb5c65fe"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from om import base, settings\n",
      "from om.components import *\n",
      "from om.data import *\n",
      "from om.util import *\n",
      "\n",
      "import pandas as pd\n",
      "import numpy as np\n",
      "import math,cobra\n",
      "\n",
      "ome = base.Session()\n",
      "model = cobra.io.load_matlab_model('/Volumes/extra/Dropbox/om_data/models/iJO1366.mat')\n",
      "\n",
      "ged = GeneExpressionData\n",
      "dged = DifferentialGeneExpressionData\n",
      "cpge = ChIPPeakGeneExpression"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stderr",
       "text": [
        "cobra/io/__init__.py:17 \u001b[1;31mUserWarning\u001b[0m: cobra.io.sbml requires libsbml\n"
       ]
      }
     ],
     "prompt_number": 1
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from IPython.display import display\n",
      "from IPython.html import widgets as W\n",
      "from IPython.utils import traitlets as T\n",
      "\n",
      "class HeatmapWidget(W.DOMWidget):\n",
      "    _view_name = T.Unicode('HeatmapView', sync=True)\n",
      "    heatmap_data = T.List(sync=True)\n",
      "    row_labels = T.List(sync=True)\n",
      "    col_labels = T.List(sync=True)\n",
      "    hcrow = T.List(sync=True)\n",
      "    hccol = T.List(sync=True)\n",
      "    minval = T.Float(sync=True)\n",
      "    maxval = T.Float(sync=True)\n",
      "    html_style = T.Unicode(sync=True)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 2
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%javascript\n",
      "\n",
      "require.config({paths: {d3: \"https://mpld3.github.io/js/d3.v3.min\"}});\n",
      "\n",
      "require([\"widgets/js/widget\", \"d3\"], function(WidgetManager, d3){\n",
      "    var HeatmapView = IPython.DOMWidgetView.extend({\n",
      "\n",
      "        render: function(){\n",
      "            \n",
      "            this.$el.append(this.model.get(\"html_style\"));\n",
      "\n",
      "            this.update();\n",
      "        },\n",
      "        update: function(){\n",
      "            var margin = { top: 175, right: 10, bottom: 150, left: 400 },\n",
      "                cellSize=18,\n",
      "                 // - margin.top - margin.bottom,\n",
      "                  //gridSize = Math.floor(width / 24),\n",
      "                legendElementWidth = cellSize*2.5,\n",
      "                colorBuckets = 9;\n",
      "                //colors = ['#005824','#1A693B','#347B53','#4F8D6B','#699F83','#83B09B','#9EC2B3','#B8D4CB','#D2E6E3','#EDF8FB','#FFFFFF','#F1EEF6','#E6D3E1','#DBB9CD','#D19EB9','#C684A4','#BB6990','#B14F7C','#A63467','#9B1A53','#91003F'];\n",
      "            var colors = [\"#081d58\",\"#253494\",\"#225ea8\",\"#1d91c0\",\"#41b6c4\",\"#7fcdbb\",\"#c7e9b4\",\"#edf8b1\",\"#ffffd9\"];\n",
      "            var rowLabel = this.model.get(\"row_labels\");\n",
      "            var colLabel = this.model.get(\"col_labels\");\n",
      "            var data = this.model.get(\"heatmap_data\");\n",
      "            var hcrow = this.model.get(\"hcrow\");\n",
      "            var hccol = this.model.get(\"hccol\");\n",
      "            \n",
      "            var col_number = colLabel.length,\n",
      "                row_number = rowLabel.length;\n",
      "            \n",
      "            var width = cellSize*col_number, // - margin.left - margin.right,\n",
      "                height = cellSize*row_number;\n",
      "            \n",
      "            var maxval = JSON.parse(this.model.get(\"maxval\"));\n",
      "            var minval = JSON.parse(this.model.get(\"minval\"));\n",
      "            \n",
      "            var colorScale = d3.scale.quantile()\n",
      "                               .domain([ minval , 0, maxval])\n",
      "                               .range(colors);\n",
      "  \n",
      "            this.svg = d3.select(this.el).append(\"svg\")\n",
      "                .attr(\"width\", width + margin.left + margin.right)\n",
      "                .attr(\"height\", height + margin.top + margin.bottom)\n",
      "                .append(\"g\")\n",
      "                .attr(\"transform\", \"translate(\" + margin.left + \",\" + margin.top + \")\");\n",
      "            \n",
      "            var svg = this.svg;\n",
      "\n",
      "            var rowSortOrder=false;\n",
      "            var colSortOrder=false;\n",
      "            var rowLabels = svg.append(\"g\")\n",
      "                  .selectAll(\".rowLabelg\")\n",
      "                  .data(rowLabel)\n",
      "                  .enter()\n",
      "                  .append(\"text\")\n",
      "                  .text(function (d) { return d; })\n",
      "                  .attr(\"x\", 0)\n",
      "                  .attr(\"y\", function (d, i) { return hcrow.indexOf(i+1) * cellSize; })\n",
      "                  .style(\"text-anchor\", \"end\")\n",
      "                  .attr(\"transform\", \"translate(-25,\" + cellSize / 1.5 + \")\")\n",
      "                  .attr(\"class\", function (d,i) { return \"rowLabel mono r\"+i;} )\n",
      "                  .on(\"mouseover\", function(d) {d3.select(this).classed(\"text-hover\",true);})\n",
      "                  .on(\"mouseout\" , function(d) {d3.select(this).classed(\"text-hover\",false);})\n",
      "                  .on(\"click\", function(d,i) {rowSortOrder=!rowSortOrder; sortbylabel(\"r\",i,rowSortOrder); this.$('#order').property(\"selectedIndex\", 4).node().focus();;});\n",
      "\n",
      "            var colLabels = svg.append(\"g\")\n",
      "                  .selectAll(\".colLabelg\")\n",
      "                  .data(colLabel)\n",
      "                  .enter()\n",
      "                  .append(\"text\")\n",
      "                  .text(function (d) { return d; })\n",
      "                  .attr(\"x\", 0)\n",
      "                  .attr(\"y\", function (d, i) { return hccol.indexOf(i+1) * cellSize; })\n",
      "                  .style(\"text-anchor\", \"left\")\n",
      "                  .attr(\"transform\", \"translate(\"+cellSize/2 + \",-25) rotate (-90)\")\n",
      "                  .attr(\"class\",  function (d,i) { return \"colLabel mono c\"+i;} )\n",
      "                  .on(\"mouseover\", function(d) {d3.select(this).classed(\"text-hover\",true);})\n",
      "                  .on(\"mouseout\" , function(d) {d3.select(this).classed(\"text-hover\",false);})\n",
      "                  .on(\"click\", function(d,i) {colSortOrder=!colSortOrder;  sortbylabel(\"c\",i,colSortOrder); this.$('#order').property(\"selectedIndex\", 4).node().focus();;});\n",
      "\n",
      "            var heatMap = svg.append(\"g\").attr(\"class\",\"g3\")\n",
      "                  .selectAll(\".cellg\")\n",
      "                  .data(data,function(d){return d.row+\":\"+d.col;})\n",
      "                  .enter()\n",
      "                  .append(\"rect\")\n",
      "                  .attr(\"x\", function(d) { return hccol.indexOf(d.col) * cellSize; })\n",
      "        .attr(\"y\", function(d) { return hcrow.indexOf(d.row) * cellSize; })\n",
      "        .attr(\"class\", function(d){return \"cell cell-border cr\"+(d.row-1)+\" cc\"+(d.col-1);})\n",
      "        .attr(\"width\", cellSize)\n",
      "        .attr(\"height\", cellSize)\n",
      "        .style(\"fill\", function(d) { return colorScale(d.value); })\n",
      "        /* .on(\"click\", function(d) {\n",
      "               var rowtext=d3.select(\".r\"+(d.row-1));\n",
      "               if(rowtext.classed(\"text-selected\")==false){\n",
      "                   rowtext.classed(\"text-selected\",true);\n",
      "               }else{\n",
      "                   rowtext.classed(\"text-selected\",false);\n",
      "               }\n",
      "        })*/\n",
      "        .on(\"mouseover\", function(d){\n",
      "               //highlight text\n",
      "               d3.select(this).classed(\"cell-hover\",true);\n",
      "               d3.selectAll(\".rowLabel\").classed(\"text-highlight\",function(r,ri){ return ri==(d.row-1);});\n",
      "               d3.selectAll(\".colLabel\").classed(\"text-highlight\",function(c,ci){ return ci==(d.col-1);});\n",
      "        \n",
      "               //Update the tooltip position and value\n",
      "               d3.select(\"#tooltip\")\n",
      "                 .style(\"left\", (d3.event.pageX+10) + \"px\")\n",
      "                 .style(\"top\", (d3.event.pageY-10) + \"px\")\n",
      "                 .select(\"#value\")\n",
      "                 .text(rowLabel[d.row-1]+\",\"+colLabel[d.col-1]+\"\\ndata:\"+d.value);  \n",
      "               //Show the tooltip\n",
      "               d3.select(\"#tooltip\").classed(\"hidden\", false);\n",
      "        })\n",
      "        .on(\"mouseout\", function(){\n",
      "               d3.select(this).classed(\"cell-hover\",false);\n",
      "               d3.selectAll(\".rowLabel\").classed(\"text-highlight\",false);\n",
      "               d3.selectAll(\".colLabel\").classed(\"text-highlight\",false);\n",
      "               d3.select(\"#tooltip\").classed(\"hidden\", true);\n",
      "        })\n",
      "        ;\n",
      "\n",
      "  var legend = svg.selectAll(\".legend\")\n",
      "      .data([-10,-8,-6,-4,-2,0,2,4,6,8,10])\n",
      "      .enter().append(\"g\")\n",
      "      .attr(\"class\", \"legend\");\n",
      " \n",
      "  console.log(height+(cellSize*2));\n",
      "  legend.append(\"rect\")\n",
      "    .attr(\"x\", function(d, i) { return legendElementWidth * i; })\n",
      "    .attr(\"y\", height+(cellSize*2))\n",
      "    .attr(\"width\", legendElementWidth)\n",
      "    .attr(\"height\", cellSize)\n",
      "    .style(\"fill\", function(d, i) { return colors[i]; });\n",
      " \n",
      "  legend.append(\"text\")\n",
      "    .attr(\"class\", \"mono\")\n",
      "    .text(function(d) { return d; })\n",
      "    .attr(\"width\", legendElementWidth)\n",
      "    .attr(\"x\", function(d, i) { return legendElementWidth * i; })\n",
      "    .attr(\"y\", height + (cellSize*4));\n",
      "\n",
      "// Change ordering of cells\n",
      "\n",
      "  function sortbylabel(rORc,i,sortOrder){\n",
      "       var t = svg.transition().duration(3000);\n",
      "       var log2r=[];\n",
      "       var sorted; // sorted is zero-based index\n",
      "       d3.selectAll(\".c\"+rORc+i) \n",
      "         .filter(function(ce){\n",
      "            log2r.push(ce.value);\n",
      "          })\n",
      "       ;\n",
      "       if(rORc==\"r\"){ // sort log2ratio of a gene\n",
      "         sorted=d3.range(col_number).sort(function(a,b){ if(sortOrder){ return log2r[b]-log2r[a];}else{ return log2r[a]-log2r[b];}});\n",
      "         t.selectAll(\".cell\")\n",
      "           .attr(\"x\", function(d) { return sorted.indexOf(d.col-1) * cellSize; })\n",
      "           ;\n",
      "         t.selectAll(\".colLabel\")\n",
      "          .attr(\"y\", function (d, i) { return sorted.indexOf(i) * cellSize; })\n",
      "         ;\n",
      "       }else{ // sort log2ratio of a contrast\n",
      "         sorted=d3.range(row_number).sort(function(a,b){if(sortOrder){ return log2r[b]-log2r[a];}else{ return log2r[a]-log2r[b];}});\n",
      "         t.selectAll(\".cell\")\n",
      "           .attr(\"y\", function(d) { return sorted.indexOf(d.row-1) * cellSize; })\n",
      "           ;\n",
      "         t.selectAll(\".rowLabel\")\n",
      "          .attr(\"y\", function (d, i) { return sorted.indexOf(i) * cellSize; })\n",
      "         ;\n",
      "       }\n",
      "  }\n",
      "\n",
      "  //d3.select(\"#order\").on(\"change\",function(){\n",
      "    this.$('#order').on(\"change\",function(){\n",
      "    order(this.value);\n",
      "  });\n",
      "  \n",
      "  function order(value){\n",
      "   if(value==\"hclust\"){\n",
      "    var t = svg.transition().duration(3000);\n",
      "    t.selectAll(\".cell\")\n",
      "      .attr(\"x\", function(d) { return hccol.indexOf(d.col) * cellSize; })\n",
      "      .attr(\"y\", function(d) { return hcrow.indexOf(d.row) * cellSize; })\n",
      "      ;\n",
      "\n",
      "    t.selectAll(\".rowLabel\")\n",
      "      .attr(\"y\", function (d, i) { return hcrow.indexOf(i+1) * cellSize; })\n",
      "      ;\n",
      "\n",
      "    t.selectAll(\".colLabel\")\n",
      "      .attr(\"y\", function (d, i) { return hccol.indexOf(i+1) * cellSize; })\n",
      "      ;\n",
      "\n",
      "   }else if (value==\"probecontrast\"){\n",
      "    var t = svg.transition().duration(3000);\n",
      "    t.selectAll(\".cell\")\n",
      "      .attr(\"x\", function(d) { return (d.col - 1) * cellSize; })\n",
      "      .attr(\"y\", function(d) { return (d.row - 1) * cellSize; })\n",
      "      ;\n",
      "\n",
      "    t.selectAll(\".rowLabel\")\n",
      "      .attr(\"y\", function (d, i) { return i * cellSize; })\n",
      "      ;\n",
      "\n",
      "    t.selectAll(\".colLabel\")\n",
      "      .attr(\"y\", function (d, i) { return i * cellSize; })\n",
      "      ;\n",
      "\n",
      "   }else if (value==\"probe\"){\n",
      "    var t = svg.transition().duration(3000);\n",
      "    t.selectAll(\".cell\")\n",
      "      .attr(\"y\", function(d) { return (d.row - 1) * cellSize; })\n",
      "      ;\n",
      "\n",
      "    t.selectAll(\".rowLabel\")\n",
      "      .attr(\"y\", function (d, i) { return i * cellSize; })\n",
      "      ;\n",
      "   }else if (value==\"contrast\"){\n",
      "    var t = svg.transition().duration(3000);\n",
      "    t.selectAll(\".cell\")\n",
      "      .attr(\"x\", function(d) { return (d.col - 1) * cellSize; })\n",
      "      ;\n",
      "    t.selectAll(\".colLabel\")\n",
      "      .attr(\"y\", function (d, i) { return i * cellSize; })\n",
      "      ;\n",
      "   }\n",
      "  }\n",
      "  // \n",
      "  var sa=d3.select(\".g3\")\n",
      "      .on(\"mousedown\", function() {\n",
      "          if( !d3.event.altKey) {\n",
      "             d3.selectAll(\".cell-selected\").classed(\"cell-selected\",false);\n",
      "             d3.selectAll(\".rowLabel\").classed(\"text-selected\",false);\n",
      "             d3.selectAll(\".colLabel\").classed(\"text-selected\",false);\n",
      "          }\n",
      "         var p = d3.mouse(this);\n",
      "         sa.append(\"rect\")\n",
      "         .attr({\n",
      "             rx      : 0,\n",
      "             ry      : 0,\n",
      "             class   : \"selection\",\n",
      "             x       : p[0],\n",
      "             y       : p[1],\n",
      "             width   : 1,\n",
      "             height  : 1\n",
      "         })\n",
      "      })\n",
      "      .on(\"mousemove\", function() {\n",
      "         var s = sa.select(\"rect.selection\");\n",
      "      \n",
      "         if(!s.empty()) {\n",
      "             var p = d3.mouse(this),\n",
      "                 d = {\n",
      "                     x       : parseInt(s.attr(\"x\"), 10),\n",
      "                     y       : parseInt(s.attr(\"y\"), 10),\n",
      "                     width   : parseInt(s.attr(\"width\"), 10),\n",
      "                     height  : parseInt(s.attr(\"height\"), 10)\n",
      "                 },\n",
      "                 move = {\n",
      "                     x : p[0] - d.x,\n",
      "                     y : p[1] - d.y\n",
      "                 }\n",
      "             ;\n",
      "      \n",
      "             if(move.x < 1 || (move.x*2<d.width)) {\n",
      "                 d.x = p[0];\n",
      "                 d.width -= move.x;\n",
      "             } else {\n",
      "                 d.width = move.x;       \n",
      "             }\n",
      "      \n",
      "             if(move.y < 1 || (move.y*2<d.height)) {\n",
      "                 d.y = p[1];\n",
      "                 d.height -= move.y;\n",
      "             } else {\n",
      "                 d.height = move.y;       \n",
      "             }\n",
      "             s.attr(d);\n",
      "      \n",
      "                 // deselect all temporary selected state objects\n",
      "             d3.selectAll('.cell-selection.cell-selected').classed(\"cell-selected\", false);\n",
      "             d3.selectAll(\".text-selection.text-selected\").classed(\"text-selected\",false);\n",
      "\n",
      "             d3.selectAll('.cell').filter(function(cell_d, i) {\n",
      "                 if(\n",
      "                     !d3.select(this).classed(\"cell-selected\") && \n",
      "                         // inner circle inside selection frame\n",
      "                     (this.x.baseVal.value)+cellSize >= d.x && (this.x.baseVal.value)<=d.x+d.width && \n",
      "                     (this.y.baseVal.value)+cellSize >= d.y && (this.y.baseVal.value)<=d.y+d.height\n",
      "                 ) {\n",
      "      \n",
      "                     d3.select(this)\n",
      "                     .classed(\"cell-selection\", true)\n",
      "                     .classed(\"cell-selected\", true);\n",
      "\n",
      "                     d3.select(\".r\"+(cell_d.row-1))\n",
      "                     .classed(\"text-selection\",true)\n",
      "                     .classed(\"text-selected\",true);\n",
      "\n",
      "                     d3.select(\".c\"+(cell_d.col-1))\n",
      "                     .classed(\"text-selection\",true)\n",
      "                     .classed(\"text-selected\",true);\n",
      "                 }\n",
      "             });\n",
      "         }\n",
      "      })\n",
      "      .on(\"mouseup\", function() {\n",
      "            // remove selection frame\n",
      "         sa.selectAll(\"rect.selection\").remove();\n",
      "      \n",
      "             // remove temporary selection marker class\n",
      "         d3.selectAll('.cell-selection').classed(\"cell-selection\", false);\n",
      "         d3.selectAll(\".text-selection\").classed(\"text-selection\",false);\n",
      "      })\n",
      "      .on(\"mouseout\", function() {\n",
      "         if(d3.event.relatedTarget.tagName=='html') {\n",
      "                 // remove selection frame\n",
      "             sa.selectAll(\"rect.selection\").remove();\n",
      "                 // remove temporary selection marker class\n",
      "             d3.selectAll('.cell-selection').classed(\"cell-selection\", false);\n",
      "             d3.selectAll(\".rowLabel\").classed(\"text-selected\",false);\n",
      "             d3.selectAll(\".colLabel\").classed(\"text-selected\",false);\n",
      "         }\n",
      "      });\n",
      "\n",
      "            \n",
      "        }\n",
      "    });\n",
      "    WidgetManager.register_widget_view(\"HeatmapView\", HeatmapView);\n",
      "});"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "javascript": [
        "\n",
        "require.config({paths: {d3: \"https://mpld3.github.io/js/d3.v3.min\"}});\n",
        "\n",
        "require([\"widgets/js/widget\", \"d3\"], function(WidgetManager, d3){\n",
        "    var HeatmapView = IPython.DOMWidgetView.extend({\n",
        "\n",
        "        render: function(){\n",
        "            \n",
        "            this.$el.append(this.model.get(\"html_style\"));\n",
        "\n",
        "            this.update();\n",
        "        },\n",
        "        update: function(){\n",
        "            var margin = { top: 175, right: 10, bottom: 150, left: 400 },\n",
        "                cellSize=18,\n",
        "                 // - margin.top - margin.bottom,\n",
        "                  //gridSize = Math.floor(width / 24),\n",
        "                legendElementWidth = cellSize*2.5,\n",
        "                colorBuckets = 9;\n",
        "                //colors = ['#005824','#1A693B','#347B53','#4F8D6B','#699F83','#83B09B','#9EC2B3','#B8D4CB','#D2E6E3','#EDF8FB','#FFFFFF','#F1EEF6','#E6D3E1','#DBB9CD','#D19EB9','#C684A4','#BB6990','#B14F7C','#A63467','#9B1A53','#91003F'];\n",
        "            var colors = [\"#081d58\",\"#253494\",\"#225ea8\",\"#1d91c0\",\"#41b6c4\",\"#7fcdbb\",\"#c7e9b4\",\"#edf8b1\",\"#ffffd9\"];\n",
        "            var rowLabel = this.model.get(\"row_labels\");\n",
        "            var colLabel = this.model.get(\"col_labels\");\n",
        "            var data = this.model.get(\"heatmap_data\");\n",
        "            var hcrow = this.model.get(\"hcrow\");\n",
        "            var hccol = this.model.get(\"hccol\");\n",
        "            \n",
        "            var col_number = colLabel.length,\n",
        "                row_number = rowLabel.length;\n",
        "            \n",
        "            var width = cellSize*col_number, // - margin.left - margin.right,\n",
        "                height = cellSize*row_number;\n",
        "            \n",
        "            var maxval = JSON.parse(this.model.get(\"maxval\"));\n",
        "            var minval = JSON.parse(this.model.get(\"minval\"));\n",
        "            \n",
        "            var colorScale = d3.scale.quantile()\n",
        "                               .domain([ minval , 0, maxval])\n",
        "                               .range(colors);\n",
        "  \n",
        "            this.svg = d3.select(this.el).append(\"svg\")\n",
        "                .attr(\"width\", width + margin.left + margin.right)\n",
        "                .attr(\"height\", height + margin.top + margin.bottom)\n",
        "                .append(\"g\")\n",
        "                .attr(\"transform\", \"translate(\" + margin.left + \",\" + margin.top + \")\");\n",
        "            \n",
        "            var svg = this.svg;\n",
        "\n",
        "            var rowSortOrder=false;\n",
        "            var colSortOrder=false;\n",
        "            var rowLabels = svg.append(\"g\")\n",
        "                  .selectAll(\".rowLabelg\")\n",
        "                  .data(rowLabel)\n",
        "                  .enter()\n",
        "                  .append(\"text\")\n",
        "                  .text(function (d) { return d; })\n",
        "                  .attr(\"x\", 0)\n",
        "                  .attr(\"y\", function (d, i) { return hcrow.indexOf(i+1) * cellSize; })\n",
        "                  .style(\"text-anchor\", \"end\")\n",
        "                  .attr(\"transform\", \"translate(-25,\" + cellSize / 1.5 + \")\")\n",
        "                  .attr(\"class\", function (d,i) { return \"rowLabel mono r\"+i;} )\n",
        "                  .on(\"mouseover\", function(d) {d3.select(this).classed(\"text-hover\",true);})\n",
        "                  .on(\"mouseout\" , function(d) {d3.select(this).classed(\"text-hover\",false);})\n",
        "                  .on(\"click\", function(d,i) {rowSortOrder=!rowSortOrder; sortbylabel(\"r\",i,rowSortOrder); this.$('#order').property(\"selectedIndex\", 4).node().focus();;});\n",
        "\n",
        "            var colLabels = svg.append(\"g\")\n",
        "                  .selectAll(\".colLabelg\")\n",
        "                  .data(colLabel)\n",
        "                  .enter()\n",
        "                  .append(\"text\")\n",
        "                  .text(function (d) { return d; })\n",
        "                  .attr(\"x\", 0)\n",
        "                  .attr(\"y\", function (d, i) { return hccol.indexOf(i+1) * cellSize; })\n",
        "                  .style(\"text-anchor\", \"left\")\n",
        "                  .attr(\"transform\", \"translate(\"+cellSize/2 + \",-25) rotate (-90)\")\n",
        "                  .attr(\"class\",  function (d,i) { return \"colLabel mono c\"+i;} )\n",
        "                  .on(\"mouseover\", function(d) {d3.select(this).classed(\"text-hover\",true);})\n",
        "                  .on(\"mouseout\" , function(d) {d3.select(this).classed(\"text-hover\",false);})\n",
        "                  .on(\"click\", function(d,i) {colSortOrder=!colSortOrder;  sortbylabel(\"c\",i,colSortOrder); this.$('#order').property(\"selectedIndex\", 4).node().focus();;});\n",
        "\n",
        "            var heatMap = svg.append(\"g\").attr(\"class\",\"g3\")\n",
        "                  .selectAll(\".cellg\")\n",
        "                  .data(data,function(d){return d.row+\":\"+d.col;})\n",
        "                  .enter()\n",
        "                  .append(\"rect\")\n",
        "                  .attr(\"x\", function(d) { return hccol.indexOf(d.col) * cellSize; })\n",
        "        .attr(\"y\", function(d) { return hcrow.indexOf(d.row) * cellSize; })\n",
        "        .attr(\"class\", function(d){return \"cell cell-border cr\"+(d.row-1)+\" cc\"+(d.col-1);})\n",
        "        .attr(\"width\", cellSize)\n",
        "        .attr(\"height\", cellSize)\n",
        "        .style(\"fill\", function(d) { return colorScale(d.value); })\n",
        "        /* .on(\"click\", function(d) {\n",
        "               var rowtext=d3.select(\".r\"+(d.row-1));\n",
        "               if(rowtext.classed(\"text-selected\")==false){\n",
        "                   rowtext.classed(\"text-selected\",true);\n",
        "               }else{\n",
        "                   rowtext.classed(\"text-selected\",false);\n",
        "               }\n",
        "        })*/\n",
        "        .on(\"mouseover\", function(d){\n",
        "               //highlight text\n",
        "               d3.select(this).classed(\"cell-hover\",true);\n",
        "               d3.selectAll(\".rowLabel\").classed(\"text-highlight\",function(r,ri){ return ri==(d.row-1);});\n",
        "               d3.selectAll(\".colLabel\").classed(\"text-highlight\",function(c,ci){ return ci==(d.col-1);});\n",
        "        \n",
        "               //Update the tooltip position and value\n",
        "               d3.select(\"#tooltip\")\n",
        "                 .style(\"left\", (d3.event.pageX+10) + \"px\")\n",
        "                 .style(\"top\", (d3.event.pageY-10) + \"px\")\n",
        "                 .select(\"#value\")\n",
        "                 .text(rowLabel[d.row-1]+\",\"+colLabel[d.col-1]+\"\\ndata:\"+d.value);  \n",
        "               //Show the tooltip\n",
        "               d3.select(\"#tooltip\").classed(\"hidden\", false);\n",
        "        })\n",
        "        .on(\"mouseout\", function(){\n",
        "               d3.select(this).classed(\"cell-hover\",false);\n",
        "               d3.selectAll(\".rowLabel\").classed(\"text-highlight\",false);\n",
        "               d3.selectAll(\".colLabel\").classed(\"text-highlight\",false);\n",
        "               d3.select(\"#tooltip\").classed(\"hidden\", true);\n",
        "        })\n",
        "        ;\n",
        "\n",
        "  var legend = svg.selectAll(\".legend\")\n",
        "      .data([-10,-8,-6,-4,-2,0,2,4,6,8,10])\n",
        "      .enter().append(\"g\")\n",
        "      .attr(\"class\", \"legend\");\n",
        " \n",
        "  console.log(height+(cellSize*2));\n",
        "  legend.append(\"rect\")\n",
        "    .attr(\"x\", function(d, i) { return legendElementWidth * i; })\n",
        "    .attr(\"y\", height+(cellSize*2))\n",
        "    .attr(\"width\", legendElementWidth)\n",
        "    .attr(\"height\", cellSize)\n",
        "    .style(\"fill\", function(d, i) { return colors[i]; });\n",
        " \n",
        "  legend.append(\"text\")\n",
        "    .attr(\"class\", \"mono\")\n",
        "    .text(function(d) { return d; })\n",
        "    .attr(\"width\", legendElementWidth)\n",
        "    .attr(\"x\", function(d, i) { return legendElementWidth * i; })\n",
        "    .attr(\"y\", height + (cellSize*4));\n",
        "\n",
        "// Change ordering of cells\n",
        "\n",
        "  function sortbylabel(rORc,i,sortOrder){\n",
        "       var t = svg.transition().duration(3000);\n",
        "       var log2r=[];\n",
        "       var sorted; // sorted is zero-based index\n",
        "       d3.selectAll(\".c\"+rORc+i) \n",
        "         .filter(function(ce){\n",
        "            log2r.push(ce.value);\n",
        "          })\n",
        "       ;\n",
        "       if(rORc==\"r\"){ // sort log2ratio of a gene\n",
        "         sorted=d3.range(col_number).sort(function(a,b){ if(sortOrder){ return log2r[b]-log2r[a];}else{ return log2r[a]-log2r[b];}});\n",
        "         t.selectAll(\".cell\")\n",
        "           .attr(\"x\", function(d) { return sorted.indexOf(d.col-1) * cellSize; })\n",
        "           ;\n",
        "         t.selectAll(\".colLabel\")\n",
        "          .attr(\"y\", function (d, i) { return sorted.indexOf(i) * cellSize; })\n",
        "         ;\n",
        "       }else{ // sort log2ratio of a contrast\n",
        "         sorted=d3.range(row_number).sort(function(a,b){if(sortOrder){ return log2r[b]-log2r[a];}else{ return log2r[a]-log2r[b];}});\n",
        "         t.selectAll(\".cell\")\n",
        "           .attr(\"y\", function(d) { return sorted.indexOf(d.row-1) * cellSize; })\n",
        "           ;\n",
        "         t.selectAll(\".rowLabel\")\n",
        "          .attr(\"y\", function (d, i) { return sorted.indexOf(i) * cellSize; })\n",
        "         ;\n",
        "       }\n",
        "  }\n",
        "\n",
        "  //d3.select(\"#order\").on(\"change\",function(){\n",
        "    this.$('#order').on(\"change\",function(){\n",
        "    order(this.value);\n",
        "  });\n",
        "  \n",
        "  function order(value){\n",
        "   if(value==\"hclust\"){\n",
        "    var t = svg.transition().duration(3000);\n",
        "    t.selectAll(\".cell\")\n",
        "      .attr(\"x\", function(d) { return hccol.indexOf(d.col) * cellSize; })\n",
        "      .attr(\"y\", function(d) { return hcrow.indexOf(d.row) * cellSize; })\n",
        "      ;\n",
        "\n",
        "    t.selectAll(\".rowLabel\")\n",
        "      .attr(\"y\", function (d, i) { return hcrow.indexOf(i+1) * cellSize; })\n",
        "      ;\n",
        "\n",
        "    t.selectAll(\".colLabel\")\n",
        "      .attr(\"y\", function (d, i) { return hccol.indexOf(i+1) * cellSize; })\n",
        "      ;\n",
        "\n",
        "   }else if (value==\"probecontrast\"){\n",
        "    var t = svg.transition().duration(3000);\n",
        "    t.selectAll(\".cell\")\n",
        "      .attr(\"x\", function(d) { return (d.col - 1) * cellSize; })\n",
        "      .attr(\"y\", function(d) { return (d.row - 1) * cellSize; })\n",
        "      ;\n",
        "\n",
        "    t.selectAll(\".rowLabel\")\n",
        "      .attr(\"y\", function (d, i) { return i * cellSize; })\n",
        "      ;\n",
        "\n",
        "    t.selectAll(\".colLabel\")\n",
        "      .attr(\"y\", function (d, i) { return i * cellSize; })\n",
        "      ;\n",
        "\n",
        "   }else if (value==\"probe\"){\n",
        "    var t = svg.transition().duration(3000);\n",
        "    t.selectAll(\".cell\")\n",
        "      .attr(\"y\", function(d) { return (d.row - 1) * cellSize; })\n",
        "      ;\n",
        "\n",
        "    t.selectAll(\".rowLabel\")\n",
        "      .attr(\"y\", function (d, i) { return i * cellSize; })\n",
        "      ;\n",
        "   }else if (value==\"contrast\"){\n",
        "    var t = svg.transition().duration(3000);\n",
        "    t.selectAll(\".cell\")\n",
        "      .attr(\"x\", function(d) { return (d.col - 1) * cellSize; })\n",
        "      ;\n",
        "    t.selectAll(\".colLabel\")\n",
        "      .attr(\"y\", function (d, i) { return i * cellSize; })\n",
        "      ;\n",
        "   }\n",
        "  }\n",
        "  // \n",
        "  var sa=d3.select(\".g3\")\n",
        "      .on(\"mousedown\", function() {\n",
        "          if( !d3.event.altKey) {\n",
        "             d3.selectAll(\".cell-selected\").classed(\"cell-selected\",false);\n",
        "             d3.selectAll(\".rowLabel\").classed(\"text-selected\",false);\n",
        "             d3.selectAll(\".colLabel\").classed(\"text-selected\",false);\n",
        "          }\n",
        "         var p = d3.mouse(this);\n",
        "         sa.append(\"rect\")\n",
        "         .attr({\n",
        "             rx      : 0,\n",
        "             ry      : 0,\n",
        "             class   : \"selection\",\n",
        "             x       : p[0],\n",
        "             y       : p[1],\n",
        "             width   : 1,\n",
        "             height  : 1\n",
        "         })\n",
        "      })\n",
        "      .on(\"mousemove\", function() {\n",
        "         var s = sa.select(\"rect.selection\");\n",
        "      \n",
        "         if(!s.empty()) {\n",
        "             var p = d3.mouse(this),\n",
        "                 d = {\n",
        "                     x       : parseInt(s.attr(\"x\"), 10),\n",
        "                     y       : parseInt(s.attr(\"y\"), 10),\n",
        "                     width   : parseInt(s.attr(\"width\"), 10),\n",
        "                     height  : parseInt(s.attr(\"height\"), 10)\n",
        "                 },\n",
        "                 move = {\n",
        "                     x : p[0] - d.x,\n",
        "                     y : p[1] - d.y\n",
        "                 }\n",
        "             ;\n",
        "      \n",
        "             if(move.x < 1 || (move.x*2<d.width)) {\n",
        "                 d.x = p[0];\n",
        "                 d.width -= move.x;\n",
        "             } else {\n",
        "                 d.width = move.x;       \n",
        "             }\n",
        "      \n",
        "             if(move.y < 1 || (move.y*2<d.height)) {\n",
        "                 d.y = p[1];\n",
        "                 d.height -= move.y;\n",
        "             } else {\n",
        "                 d.height = move.y;       \n",
        "             }\n",
        "             s.attr(d);\n",
        "      \n",
        "                 // deselect all temporary selected state objects\n",
        "             d3.selectAll('.cell-selection.cell-selected').classed(\"cell-selected\", false);\n",
        "             d3.selectAll(\".text-selection.text-selected\").classed(\"text-selected\",false);\n",
        "\n",
        "             d3.selectAll('.cell').filter(function(cell_d, i) {\n",
        "                 if(\n",
        "                     !d3.select(this).classed(\"cell-selected\") && \n",
        "                         // inner circle inside selection frame\n",
        "                     (this.x.baseVal.value)+cellSize >= d.x && (this.x.baseVal.value)<=d.x+d.width && \n",
        "                     (this.y.baseVal.value)+cellSize >= d.y && (this.y.baseVal.value)<=d.y+d.height\n",
        "                 ) {\n",
        "      \n",
        "                     d3.select(this)\n",
        "                     .classed(\"cell-selection\", true)\n",
        "                     .classed(\"cell-selected\", true);\n",
        "\n",
        "                     d3.select(\".r\"+(cell_d.row-1))\n",
        "                     .classed(\"text-selection\",true)\n",
        "                     .classed(\"text-selected\",true);\n",
        "\n",
        "                     d3.select(\".c\"+(cell_d.col-1))\n",
        "                     .classed(\"text-selection\",true)\n",
        "                     .classed(\"text-selected\",true);\n",
        "                 }\n",
        "             });\n",
        "         }\n",
        "      })\n",
        "      .on(\"mouseup\", function() {\n",
        "            // remove selection frame\n",
        "         sa.selectAll(\"rect.selection\").remove();\n",
        "      \n",
        "             // remove temporary selection marker class\n",
        "         d3.selectAll('.cell-selection').classed(\"cell-selection\", false);\n",
        "         d3.selectAll(\".text-selection\").classed(\"text-selection\",false);\n",
        "      })\n",
        "      .on(\"mouseout\", function() {\n",
        "         if(d3.event.relatedTarget.tagName=='html') {\n",
        "                 // remove selection frame\n",
        "             sa.selectAll(\"rect.selection\").remove();\n",
        "                 // remove temporary selection marker class\n",
        "             d3.selectAll('.cell-selection').classed(\"cell-selection\", false);\n",
        "             d3.selectAll(\".rowLabel\").classed(\"text-selected\",false);\n",
        "             d3.selectAll(\".colLabel\").classed(\"text-selected\",false);\n",
        "         }\n",
        "      });\n",
        "\n",
        "            \n",
        "        }\n",
        "    });\n",
        "    WidgetManager.register_widget_view(\"HeatmapView\", HeatmapView);\n",
        "});"
       ],
       "metadata": {},
       "output_type": "display_data",
       "text": [
        "<IPython.core.display.Javascript at 0x10a151a10>"
       ]
      }
     ],
     "prompt_number": 3
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "html_style = \"\"\"\n",
      "     <style>\n",
      "      /* disable text selection */\n",
      "      svg *::selection {\n",
      "         background : transparent;\n",
      "      }\n",
      "     \n",
      "      svg *::-moz-selection {\n",
      "         background:transparent;\n",
      "      } \n",
      "     \n",
      "      svg *::-webkit-selection {\n",
      "         background:transparent;\n",
      "      }\n",
      "      rect.selection {\n",
      "        stroke          : #333;\n",
      "        stroke-dasharray: 4px;\n",
      "        stroke-opacity  : 0.5;\n",
      "        fill            : transparent;\n",
      "      }\n",
      "\n",
      "      rect.cell-border {\n",
      "        stroke: #eee;\n",
      "        stroke-width:0.3px;   \n",
      "      }\n",
      "\n",
      "      rect.cell-selected {\n",
      "        stroke: rgb(51,102,153);\n",
      "        stroke-width:0.5px;   \n",
      "      }\n",
      "\n",
      "      rect.cell-hover {\n",
      "        stroke: #F00;\n",
      "        stroke-width:0.3px;   \n",
      "      }\n",
      "\n",
      "      text.mono {\n",
      "        font-size: 10pt;\n",
      "        font-family: Consolas, courier;\n",
      "        fill: #aaa;\n",
      "      }\n",
      "\n",
      "      text.text-selected {\n",
      "        fill: #000;\n",
      "      }\n",
      "\n",
      "      text.text-highlight {\n",
      "        fill: #c00;\n",
      "      }\n",
      "      text.text-hover {\n",
      "        fill: #00C;\n",
      "      }\n",
      "      #tooltip {\n",
      "        position: fixed;\n",
      "        width: 200px;\n",
      "        height: auto;\n",
      "        padding: 10px;\n",
      "        background-color: white;\n",
      "        -webkit-border-radius: 10px;\n",
      "        -moz-border-radius: 10px;\n",
      "        border-radius: 10px;\n",
      "        -webkit-box-shadow: 4px 4px 10px rgba(0, 0, 0, 0.4);\n",
      "        -moz-box-shadow: 4px 4px 10px rgba(0, 0, 0, 0.4);\n",
      "        box-shadow: 4px 4px 10px rgba(0, 0, 0, 0.4);\n",
      "        pointer-events: none;\n",
      "      }\n",
      "\n",
      "      #tooltip.hidden {\n",
      "        display: none;\n",
      "      }\n",
      "\n",
      "      #tooltip p {\n",
      "        margin: 0;\n",
      "        font-family: sans-serif;\n",
      "        font-size: 12px;\n",
      "        line-height: 20px;\n",
      "      }\n",
      "    </style>\n",
      "\n",
      "\n",
      "<div id=\"tooltip\" class=\"hidden\">\n",
      "        <p><span id=\"value\"></p>\n",
      "</div>\n",
      "\n",
      " \n",
      "  <select id=\"order\">\n",
      "  <option value=\"hclust\">by cluster</option>\n",
      "  <option value=\"probecontrast\">by probe name and contrast name</option>\n",
      "  <option value=\"probe\">by probe name</option>\n",
      "  <option value=\"contrast\">by contrast name</option>\n",
      "  <option value=\"custom\">by log2 ratio</option>\n",
      "  </select>\n",
      "  </select>\n",
      "<div id=\"chart\" style='overflow:auto; width:960px; height:auto;'></div>\n",
      "\"\"\""
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 4
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "ome.query(GeneGroup).all()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 5,
       "text": [
        "[Gene Group (#1, class 1) 7 genes,\n",
        " Gene Group (#5, glycerol_repressed) 58 genes,\n",
        " Gene Group (#6, glycerol_activated_dict) 106 genes,\n",
        " Gene Group (#7, fructose_repressed) 48 genes,\n",
        " Gene Group (#9, fructose_activated) 89 genes,\n",
        " Gene Group (#10, glycerol_activated) 106 genes,\n",
        " Gene Group (#18, crp_catabolic_genes) 129 genes,\n",
        " Gene Group (#19, crp_anabolic_genes) 99 genes,\n",
        " Gene Group (#20, crp_chemiosmotic_genes) 57 genes,\n",
        " Gene Group (#21, Glycolysis / Gluconeogenesis) 40 genes,\n",
        " Gene Group (#22, Citrate cycle (TCA cycle)) 28 genes,\n",
        " Gene Group (#23, Pentose phosphate pathway) 30 genes,\n",
        " Gene Group (#24, Pentose and glucuronate interconversions) 27 genes,\n",
        " Gene Group (#25, Fructose and mannose metabolism) 38 genes,\n",
        " Gene Group (#26, Galactose metabolism) 33 genes,\n",
        " Gene Group (#27, Ascorbate and aldarate metabolism) 14 genes,\n",
        " Gene Group (#28, Fatty acid biosynthesis) 12 genes,\n",
        " Gene Group (#29, Fatty acid metabolism) 16 genes,\n",
        " Gene Group (#30, Ubiquinone and other terpenoid-quinone biosynthesis) 17 genes,\n",
        " Gene Group (#31, Oxidative phosphorylation) 41 genes,\n",
        " Gene Group (#32, Purine metabolism) 87 genes,\n",
        " Gene Group (#33, Pyrimidine metabolism) 53 genes,\n",
        " Gene Group (#34, Alanine, aspartate and glutamate metabolism) 29 genes,\n",
        " Gene Group (#35, Glycine, serine and threonine metabolism) 32 genes,\n",
        " Gene Group (#36, Cysteine and methionine metabolism) 28 genes,\n",
        " Gene Group (#37, Valine, leucine and isoleucine degradation) 12 genes,\n",
        " Gene Group (#38, Geraniol degradation) 6 genes,\n",
        " Gene Group (#39, Valine, leucine and isoleucine biosynthesis) 21 genes,\n",
        " Gene Group (#40, Lysine biosynthesis) 16 genes,\n",
        " Gene Group (#41, Lysine degradation) 11 genes,\n",
        " Gene Group (#42, Arginine and proline metabolism) 43 genes,\n",
        " Gene Group (#43, Histidine metabolism) 9 genes,\n",
        " Gene Group (#44, Tyrosine metabolism) 9 genes,\n",
        " Gene Group (#45, Phenylalanine metabolism) 18 genes,\n",
        " Gene Group (#46, gamma-Hexachlorocyclohexane degradation) 5 genes,\n",
        " Gene Group (#47, Benzoate degradation via hydroxylation) 5 genes,\n",
        " Gene Group (#48, Fluorobenzoate degradation) 1 genes,\n",
        " Gene Group (#49, Tryptophan metabolism) 11 genes,\n",
        " Gene Group (#50, Phenylalanine, tyrosine and tryptophan biosynthesis) 21 genes,\n",
        " Gene Group (#51, Novobiocin biosynthesis) 4 genes,\n",
        " Gene Group (#52, beta-Alanine metabolism) 15 genes,\n",
        " Gene Group (#53, Taurine and hypotaurine metabolism) 6 genes,\n",
        " Gene Group (#54, Phosphonate and phosphinate metabolism) 2 genes,\n",
        " Gene Group (#55, Selenoamino acid metabolism) 14 genes,\n",
        " Gene Group (#56, Cyanoamino acid metabolism) 6 genes,\n",
        " Gene Group (#57, D-Glutamine and D-glutamate metabolism) 5 genes,\n",
        " Gene Group (#58, D-Alanine metabolism) 4 genes,\n",
        " Gene Group (#59, Glutathione metabolism) 17 genes,\n",
        " Gene Group (#60, Starch and sucrose metabolism) 32 genes,\n",
        " Gene Group (#61, Other glycan degradation) 4 genes,\n",
        " Gene Group (#62, Amino sugar and nucleotide sugar metabolism) 44 genes,\n",
        " Gene Group (#63, Streptomycin biosynthesis) 9 genes,\n",
        " Gene Group (#64, Polyketide sugar unit biosynthesis) 6 genes,\n",
        " Gene Group (#65, Lipopolysaccharide biosynthesis) 27 genes,\n",
        " Gene Group (#66, Peptidoglycan biosynthesis) 23 genes,\n",
        " Gene Group (#67, Glycerolipid metabolism) 13 genes,\n",
        " Gene Group (#68, Inositol phosphate metabolism) 3 genes,\n",
        " Gene Group (#69, Glycerophospholipid metabolism) 28 genes,\n",
        " Gene Group (#70, Arachidonic acid metabolism) 2 genes,\n",
        " Gene Group (#71, alpha-Linolenic acid metabolism) 3 genes,\n",
        " Gene Group (#72, Sphingolipid metabolism) 3 genes,\n",
        " Gene Group (#73, Pyruvate metabolism) 43 genes,\n",
        " Gene Group (#74, Biphenyl degradation) 2 genes,\n",
        " Gene Group (#75, Toluene and xylene degradation) 2 genes,\n",
        " Gene Group (#76, 1- and 2-Methylnaphthalene degradation) 4 genes,\n",
        " Gene Group (#77, Naphthalene and anthracene degradation) 1 genes,\n",
        " Gene Group (#78, 1,4-Dichlorobenzene degradation) 3 genes,\n",
        " Gene Group (#79, Fluorene degradation) 3 genes,\n",
        " Gene Group (#80, Carbazole degradation) 2 genes,\n",
        " Gene Group (#81, Glyoxylate and dicarboxylate metabolism) 36 genes,\n",
        " Gene Group (#82, Benzoate degradation via CoA ligation) 16 genes,\n",
        " Gene Group (#83, Trinitrotoluene degradation) 7 genes,\n",
        " Gene Group (#84, Propanoate metabolism) 30 genes,\n",
        " Gene Group (#85, 3-Chloroacrylic acid degradation) 4 genes,\n",
        " Gene Group (#86, Ethylbenzene degradation) 4 genes,\n",
        " Gene Group (#87, Styrene degradation) 3 genes,\n",
        " Gene Group (#88, Butanoate metabolism) 35 genes,\n",
        " Gene Group (#89, C5-Branched dibasic acid metabolism) 9 genes,\n",
        " Gene Group (#90, One carbon pool by folate) 13 genes,\n",
        " Gene Group (#91, Methane metabolism) 15 genes,\n",
        " Gene Group (#92, Thiamine metabolism) 14 genes,\n",
        " Gene Group (#93, Riboflavin metabolism) 11 genes,\n",
        " Gene Group (#94, Vitamin B6 metabolism) 9 genes,\n",
        " Gene Group (#95, Nicotinate and nicotinamide metabolism) 17 genes,\n",
        " Gene Group (#96, Pantothenate and CoA biosynthesis) 21 genes,\n",
        " Gene Group (#97, Biotin metabolism) 7 genes,\n",
        " Gene Group (#98, Lipoic acid metabolism) 3 genes,\n",
        " Gene Group (#99, Folate biosynthesis) 12 genes,\n",
        " Gene Group (#100, Porphyrin and chlorophyll metabolism) 25 genes,\n",
        " Gene Group (#101, Terpenoid backbone biosynthesis) 13 genes,\n",
        " Gene Group (#102, Limonene and pinene degradation) 5 genes,\n",
        " Gene Group (#103, Nitrogen metabolism) 35 genes,\n",
        " Gene Group (#104, Sulfur metabolism) 13 genes,\n",
        " Gene Group (#105, Caprolactam degradation) 4 genes,\n",
        " Gene Group (#106, Aminoacyl-tRNA biosynthesis) 26 genes,\n",
        " Gene Group (#107, Biosynthesis of unsaturated fatty acids) 6 genes,\n",
        " Gene Group (#108, Biosynthesis of siderophore group nonribosomal peptides) 6 genes,\n",
        " Gene Group (#109, ABC transporters) 182 genes,\n",
        " Gene Group (#110, Two-component system) 130 genes,\n",
        " Gene Group (#111, Bacterial chemotaxis) 21 genes,\n",
        " Gene Group (#112, Flagellar assembly) 38 genes,\n",
        " Gene Group (#113, Phosphotransferase system (PTS)) 45 genes,\n",
        " Gene Group (#114, Ribosome) 79 genes,\n",
        " Gene Group (#115, RNA degradation) 15 genes,\n",
        " Gene Group (#116, RNA polymerase) 4 genes,\n",
        " Gene Group (#117, DNA replication) 17 genes,\n",
        " Gene Group (#118, Protein export) 19 genes,\n",
        " Gene Group (#119, Bacterial secretion system) 32 genes,\n",
        " Gene Group (#120, Base excision repair) 14 genes,\n",
        " Gene Group (#121, Nucleotide excision repair) 8 genes,\n",
        " Gene Group (#122, Mismatch repair) 22 genes,\n",
        " Gene Group (#123, Homologous recombination) 27 genes,\n",
        " Gene Group (#124, zero_non_rpod) 16 genes]"
       ]
      }
     ],
     "prompt_number": 5
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "gene_list =  ome.query(GeneGroup).filter(GeneGroup.name.ilike('DNA%')).first().genes\n",
      "    \n",
      "analysis_type = GeneExpressionData\n",
      "dataset_type = 'rna_seq_experiment'\n",
      "\n",
      "\n",
      "\n",
      "all_data = ome.query(analysis_type).filter(and_(analysis_type.gene_name.in_([g.name for g in gene_list]),\n",
      "                                              analysis_type.dataset_type.ilike(dataset_type))).all()\n",
      "    \n",
      "datasets = set(['_'.join([x.dataset_type,x.strain,x.carbon_source,x.nitrogen_source,x.electron_acceptor]) for x in all_data])\n",
      "\n",
      "genes_data = pd.DataFrame(index=[g.name for g in gene_list], columns=datasets)\n",
      "\n",
      "for x in all_data:\n",
      "    genes_data.ix[x.gene_name]['_'.join([x.dataset_type,x.strain,x.carbon_source,x.nitrogen_source,x.electron_acceptor])] = x.value\n",
      "\n",
      "\n",
      "heatmap_data = []\n",
      "for i,g in enumerate(genes_data.index):\n",
      "    for j,c in enumerate(genes_data.columns):\n",
      "        heatmap_data.append({\"row\": j+1, \"col\": i+1, \"value\": np.log(genes_data.ix[g][c])})"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 8
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def gene_heatmap(gene_list, analysis_type=GeneExpressionData, dataset_type='%', \n",
      "                                                              strain1=ome.query(Strain).all(), strain2=None,\n",
      "                                                              environments1=ome.query(InVivoEnvironment).all(), environments2=None):\n",
      "\n",
      "    dataset_id_map = {}\n",
      "    if analysis_type == GeneExpressionData:\n",
      "        all_data = ome.query(analysis_type).filter(and_(analysis_type.gene_name.in_([g.name for g in gene_list]),\n",
      "                                                        analysis_type.dataset_type.ilike(dataset_type))).all()\n",
      "    \n",
      "        datasets = set(['_'.join([x.dataset_type,x.strain,x.carbon_source,x.nitrogen_source,x.electron_acceptor]) for x in all_data])\n",
      "        \n",
      "                \n",
      "        genes_data = pd.DataFrame(index=[g.name for g in gene_list], columns=datasets)\n",
      "\n",
      "        for x in all_data:\n",
      "            genes_data.ix[x.gene_name]['_'.join([x.dataset_type,x.strain,x.carbon_source,x.nitrogen_source,x.electron_acceptor])] = x.value\n",
      "\n",
      "        \n",
      "    elif analysis_type == DifferentialGeneExpressionData:\n",
      "        all_data = ome.query(analysis_type).\\\n",
      "                              filter(and_(analysis_type.gene_id.in_([g.id for g in gene_list]),\n",
      "                                          analysis_type.strain1.in_([x.name for x in strain1]),\n",
      "                                          analysis_type.strain2.in_([x.name for x in strain2]),\n",
      "                                          analysis_type.environment_id_1.in_([x.id for x in environments1]),\n",
      "                                          analysis_type.environment_id_2.in_([x.id for x in environments2]))).all()\n",
      "        \n",
      "        datasets = set(['_'.join([x.strain1+'/'+x.strain2,x.carbon_source1+'/'+x.carbon_source2,x.nitrogen_source1,x.electron_acceptor1]) for x in all_data])\n",
      "        genes_data = pd.DataFrame(index=[g.name for g in gene_list], columns=datasets)\n",
      "\n",
      "        for x in all_data:\n",
      "            genes_data.ix[x.gene_name]['_'.join([x.strain1+'/'+x.strain2,x.carbon_source1+'/'+x.carbon_source2,x.nitrogen_source1,x.electron_acceptor1])] = x.value\n",
      "\n",
      "        \n",
      "    elif analysis_type == ChIPPeakGeneExpression:\n",
      "        all_data = ome.query(analysis_type).\\\n",
      "                              filter(and_(analysis_type.gene_id.in_([g.id for g in gene_list]),\n",
      "                                          analysis_type.strain1.in_([x.name for x in strain1]),\n",
      "                                          analysis_type.strain2.in_([x.name for x in strain2]),\n",
      "                                          analysis_type.environment_id.in_([x.id for x in environments2]))).all()\n",
      "        \n",
      "        datasets = set(['_'.join([x.strain1,x.carbon_source,x.nitrogen_source,x.electron_acceptor]) for x in all_data])\n",
      "        genes_data = pd.DataFrame(index=[g.name for g in gene_list], columns=datasets)\n",
      "\n",
      "        for x in all_data:\n",
      "            genes_data.ix[x.gene_name]['_'.join([x.strain1,x.carbon_source,x.nitrogen_source,x.electron_acceptor])] = x.value\n",
      "\n",
      "            \n",
      "    \n",
      "    genes_data = genes_data.dropna(how='all').fillna(0.)\n",
      "    col_labels = list(genes_data.index)\n",
      "    row_labels = list(datasets)\n",
      "    \n",
      "    \n",
      "    heatmap_data = []\n",
      "    for i,g in enumerate(genes_data.index):\n",
      "        for j,c in enumerate(genes_data.columns):\n",
      "            if genes_data.ix[g][c] == -float('Inf'): value = -10\n",
      "            elif genes_data.ix[g][c] == float('Inf'): value = 10\n",
      "            else: value = genes_data.ix[g][c]\n",
      "            heatmap_data.append({\"row\": j+1, \"col\": i+1, \"value\": value})\n",
      "    \n",
      "    hcrow = range(len(row_labels))\n",
      "    hccol = range(len(col_labels))\n",
      "    np.random.shuffle(hcrow)\n",
      "    np.random.shuffle(hccol)\n",
      "    \n",
      "    return hcrow,hccol,row_labels,col_labels,heatmap_data\n",
      "    #return {'hcrow': hcrow, 'hccol': hccol, 'row_labels':row_labels, \n",
      "    #                                        'col_labels':col_labels, \n",
      "    #                                        'heatmap_data':heatmap_data,\n",
      "    #                                        'maxval' : max([x['value'] for x in heatmap_data]),\n",
      "    #                                        'minval' : min([x['value'] for x in heatmap_data])}"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 9
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "catabolic_genes = ome.query(GeneGroup).filter(GeneGroup.name.ilike('crp_cata%')).first().genes\n",
      "hcrow,hccol,row_labels,col_labels,heatmap_data = gene_heatmap(catabolic_genes, \n",
      "                                                              analysis_type=ChIPPeakGeneExpression,\n",
      "                                                              strain1=ome.query(Strain).filter(Strain.name.in_(['delta-crp','delAr1','delAr2','delAr1delAr2'])),\n",
      "                                                              strain2=ome.query(Strain).all(),\n",
      "                                                              environments2=ome.query(InVivoEnvironment).\\\n",
      "                                                                                filter(InVivoEnvironment.carbon_source.in_(['glycerol','fructose','glucose'])).all())\n",
      "\n",
      "\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 10
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "catabolic_heatmap_diagram = HeatmapWidget(html_style=html_style,\n",
      "                                row_labels=row_labels, \n",
      "                                col_labels=col_labels,\n",
      "                                hcrow=hcrow,\n",
      "                                hccol=hccol,\n",
      "                                minval= min([x['value'] for x in heatmap_data]),\n",
      "                                maxval= max([x['value'] for x in heatmap_data]),\n",
      "                                heatmap_data=heatmap_data)\n",
      "display(catabolic_heatmap_diagram)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 11
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "anabolic_genes = ome.query(GeneGroup).filter(GeneGroup.name.ilike('crp_anab%')).first().genes\n",
      "hcrow,hccol,row_labels,col_labels,heatmap_data = gene_heatmap(anabolic_genes, \n",
      "                                                              analysis_type=ChIPPeakGeneExpression,\n",
      "                                                              strain1=ome.query(Strain).filter(Strain.name.in_(['delta-crp','delAr1','delAr2','delAr1delAr2'])),\n",
      "                                                              strain2=ome.query(Strain).all(),\n",
      "                                                              environments2=ome.query(InVivoEnvironment).\\\n",
      "                                                                                filter(InVivoEnvironment.carbon_source.in_(['glycerol','fructose','glucose'])).all())\n",
      "\n",
      "\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 8
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "anabolic_heatmap_diagram = HeatmapWidget(html_style=html_style,\n",
      "                                row_labels=row_labels, \n",
      "                                col_labels=col_labels,\n",
      "                                hcrow=hcrow,\n",
      "                                hccol=hccol,\n",
      "                                minval= min([x['value'] for x in heatmap_data]),\n",
      "                                maxval= max([x['value'] for x in heatmap_data]),\n",
      "                                heatmap_data=heatmap_data)\n",
      "display(anabolic_heatmap_diagram)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 9
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "chemiosmotic_genes = ome.query(GeneGroup).filter(GeneGroup.name.ilike('crp_chemi%')).first().genes\n",
      "hcrow,hccol,row_labels,col_labels,heatmap_data = gene_heatmap(chemiosmotic_genes, \n",
      "                                                              analysis_type=ChIPPeakGeneExpression,\n",
      "                                                              strain1=ome.query(Strain).filter(Strain.name.in_(['delta-crp','delAr1','delAr2','delAr1delAr2'])),\n",
      "                                                              strain2=ome.query(Strain).all(),\n",
      "                                                              environments2=ome.query(InVivoEnvironment).\\\n",
      "                                                                                filter(InVivoEnvironment.carbon_source.in_(['glycerol','fructose','glucose'])).all())\n",
      "\n",
      "\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 10
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "chemiosmotic_heatmap_diagram = HeatmapWidget(html_style=html_style,\n",
      "                                row_labels=row_labels, \n",
      "                                col_labels=col_labels,\n",
      "                                hcrow=hcrow,\n",
      "                                hccol=hccol,\n",
      "                                minval= min([x['value'] for x in heatmap_data]),\n",
      "                                maxval= max([x['value'] for x in heatmap_data]),\n",
      "                                heatmap_data=heatmap_data)\n",
      "display(chemiosmotic_heatmap_diagram)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 11
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}